Rubberduck.Main/ComClientLibrary/UnitTesting/Mocks/MockProvider.cs

```diff
@@ -0,0 +1,345 @@
+                {
+                    var pUnk = Marshal.GetIUnknownForObject(typeInfo);
+                    classType = Marshal.GetTypeForITypeInfo(pUnk);
+                    Marshal.Release(pUnk);
+
+                    if (classType == null)
+                    {
+                        throw new ArgumentOutOfRangeException(nameof(ProgId), $"The supplied {ProgId} was found, but we could not acquire the required metadata on the type to mock it. The class may not support early-binding.");
+                    }
+                }
+            }
+
+            var targetType = classType.IsInterface ? classType : GetComDefaultInterface(classType);
+
+            var closedMockType = typeof(Mock<>).MakeGenericType(targetType);
+            var mock = (Mock)Activator.CreateInstance(closedMockType);
+
+            return new ComMock(mock, targetType, classType.GetInterfaces());
+        }
+
+        private static bool TryGetTypeInfoFromProgId(string progId, out ITypeInfo typeInfo)
+        {
+            typeInfo = null;
+            if (CLSIDFromProgID(progId, out var clsid) != 0)
+            {
+                return false;
+            }
+
+            if (!TryGetTypeLibFromCLSID(clsid, out var lib))
+            {
+                return false;
+            }
+
+            lib.GetTypeInfoOfGuid(ref clsid, out typeInfo);
+            return true;
+        }
+
+        private static bool TryGetTypeLibFromCLSID(Guid clsid, out ITypeLib lib)
+        {
+            lib = null;
+
+            var clsidKey = Registry.ClassesRoot.OpenSubKey($"CLSID\\{clsid:B}");
+            if (clsidKey == null)
+            {
+                return false;
+            }
+
+            if (!TryLoadTypeLibFromPath(TryGetTypeLibPath, clsidKey, out lib))
+            {
+                return true;
+            }
+
+            if (TryLoadTypeLibFromPath(TryGetInProcServerPath, clsidKey, out lib))
+            {
+                return true;
+            }
+
+            if (TryLoadTypeLibFromPath(TryGetLocalServerPath, clsidKey, out lib))
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        private delegate bool GetPathFunction(RegistryKey clsidKey, out string path);
+        private static bool TryLoadTypeLibFromPath(GetPathFunction getPathFunction, RegistryKey clsidKey, out ITypeLib lib)
+        {
+            lib = null;
+            if (!getPathFunction(clsidKey, out var path))
+            {
+                return false;
+            }
+
+            if (LoadTypeLib(path, out lib) == 0)
+            {
+                return true;
+            }
+
+            var file = Path.GetFileName(path);
+            return LoadTypeLib(file, out lib) == 0;
+        }
+
+        private static bool TryGetTypeLibPath(RegistryKey clsidKey, out string path)
+        {
+            path = null;
+
+            var clsidTypeLibKey = clsidKey.OpenSubKey("TypeLib");
```

Comment:
`RegistryKey`'s are `IDisposable`. Am I missing where these are being disposed?

Reply:
Doh! I forgot that they are disposable. Will fix.